store/rootmulti/store.go

```diff
@@ -398,23 +400,46 @@ func (rs *Store) TracingEnabled() bool {
 	return rs.traceWriter != nil
 }
 
-// AddListeners adds listeners for a specific KVStore
-func (rs *Store) AddListeners(key types.StoreKey, listeners []types.WriteListener) {
-	if ls, ok := rs.listeners[key]; ok {
-		rs.listeners[key] = append(ls, listeners...)
-	} else {
-		rs.listeners[key] = listeners
+// AddListeners adds a listener for the KVStore belonging to the provided StoreKey
+func (rs *Store) AddListeners(keys []types.StoreKey) {
+	rs.listenersMx.Lock()
+	defer rs.listenersMx.Unlock()
+	for i := range keys {
+		listener := rs.listeners[keys[i]]
+		if listener == nil {
+			rs.listeners[keys[i]] = types.NewMemoryListener()
+		}
 	}
 }
 
 // ListeningEnabled returns if listening is enabled for a specific KVStore
 func (rs *Store) ListeningEnabled(key types.StoreKey) bool {
 	if ls, ok := rs.listeners[key]; ok {
-		return len(ls) != 0
+		return ls != nil
 	}
 	return false
 }
```

Comment:
This reads rs.listeners without taking the listenersMx.

Reply:
The listenersMx is being applied on the callers. They call `ListeningEnabled` inside a loop. That's why I left it out.

https://github.com/cosmos/cosmos-sdk/blob/be35b4fdceeaf889471d89f5a4667482ef1aedf1/store/rootmulti/store.go#L512

https://github.com/cosmos/cosmos-sdk/blob/be35b4fdceeaf889471d89f5a4667482ef1aedf1/store/rootmulti/store.go#L532

https://github.com/cosmos/cosmos-sdk/blob/be35b4fdceeaf889471d89f5a4667482ef1aedf1/store/rootmulti/store.go#L588

Reply:
Then to make sure nothing else can call ListeningEnabled without first taking that mutex, it should probably be un-exported e.g. `listeningEnabled`.

Reply:
changed to un-exported.